src/librustc_trans/trans/consts.rs

```diff
@@ -664,12 +662,19 @@ fn const_expr_unadjusted<'a, 'tcx>(cx: &CrateContext<'a, 'tcx>,
             adt::const_get_field(cx, &*brepr, bv, vinfo.discr, idx.node)
         },
         hir::ExprIndex(ref base, ref index) => {
+            //honor the const_indexing feature if this is in a CONST expression
+            if !cx.sess().features.borrow().const_indexing && trueconst == TrueConst::Yes {
+                let e = ConstEvalErr { span: index.span, kind: ErrKind::IndexOpFeatureGated };
+                cx.sess().span_err(index.span, &e.description());
+                return Err(ConstEvalFailure::Compiletime(e));
+            }
+
             let (bv, bt) = try!(const_expr(cx, &**base, param_substs, fn_args, trueconst));
-            let iv = match eval_const_expr_partial(cx.tcx(), &index, ExprTypeChecked, None) {
-                Ok(ConstVal::Int(i)) => i as u64,
-                Ok(ConstVal::Uint(u)) => u,
-                _ => cx.sess().span_bug(index.span,
-                                        "index is not an integer-constant expression")
+            let iv = try!(const_expr(cx, &**index, param_substs, fn_args, trueconst)).0;
```

Comment:
`trueconst` should probably be `TrueConst::Yes` (definitely change this just to be on the safe side).